### PR TITLE
Adapt remove and remove_if to C++20

### PR DESCRIPTION
**Pull request by gonidelis**
This adapts `hpx::remove` and `hpx::ranges:remove_if` to C++20 with the latest API, supporting `iterator` - `sentinel` pairs and `ranges`.

flyby: adds `make_subrange` utility function.

#### todo
- [x] We need to provide consistent tests for the `iterator` - `sentinel` cases. I think `remove` is a good starting point for that, but I have not managed to make it work yet. We could accept this without the test and work all the `iterator`-`sentinel` test cases in a separate 
- [x] Extra tests will be added soon for any other overloads (e.g. without exec-policy). ~**Please don't merge yet**~.
- [x] Provide docs for iterator-sentinel `ranges::remove` and `ranges::remove_if` overloads.

### CODE REVIEW

File: libs/parallelism/algorithms/tests/unit/algorithms/remove_tests.hpp [hunk 1]
```diff
@@ -439,6 +437,8 @@ void test_remove_etc(ExPolicy policy, IteratorTag, DataType, int rand_base,
 
         if (test_for_remove_if)
         {
+            // The new overloads do not accept projections.
+            // We use the deprecated version instead.
             result = hpx::parallel::remove_if(
```
[the remark this answers is not in the document]
Comment (msimberg) [reply libs/parallelism/algorithms/tests/unit/algorithms/remove_tests.hpp:442]: As much of a problem as it was with earlier PRs. You can ignore the deprecation warnings the same way you did for `transform`.

**Comment by gonidelis** [comment]
@hkaiser @msimberg The PR is complete. Please review (once CI is finished).

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp [hunk 2]
```diff
@@ -50,14 +252,18 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+            Iter first, Sent last, UnaryPredicate p, Proj proj)
         {
-            return std::remove_if(first, last,
-                util::invoke_projected<Pred, Proj>(
-                    std::forward<Pred>(pred), std::forward<Proj>(proj)));
+            first = hpx::parallel::v1::detail::sequential_find_if(
+                first, last, p, proj);
+            if (first != last)
+                for (Iter i = first; ++i != last;)
+                    if (!p(*i))
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp:264]: Two comments: 
- please use `hpx::util::invoke` for this
- should the result of `*i` be projected?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp [hunk 3]
```diff
@@ -354,3 +455,95 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+    {
+    private:
+        // clang-format off
+        template <typename FwdIter,
+            typename T, HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<FwdIter>::value
+            )>
+        // clang-format on
+        friend FwdIter tag_invoke(
+            hpx::remove_t, FwdIter first, FwdIter last, const T& value)
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp:521]: Minor nitpick: the HPX codebase uses the 'east const' notation, could you please adapt to this style?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp (hunk 2 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp]: Everything should be on its place now. I don't expect complains from CI either.

**Review by msimberg — DISMISSED** [review]
Looks good to me! @hkaiser any further comments?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp [hunk 4]
```diff
@@ -354,3 +456,95 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+    ///////////////////////////////////////////////////////////////////////////
+    // CPO for hpx::remove_if
+    HPX_INLINE_CONSTEXPR_VARIABLE struct remove_if_t final
+      : hpx::functional::tag<remove_if_t>
+    {
+        // clang-format off
+        template <typename FwdIter,
+            typename Pred, HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<FwdIter>::value
+            )>
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove.hpp:470]: Could we have a check for the validity of the predicate here as well, please (same for other overloads below):

```suggestion
        template <typename FwdIter, typename Pred, 
            HPX_CONCEPT_REQUIRES_(
                hpx::traits::is_iterator<FwdIter>::value &&
                hpx::is_invocable_v<Pred,
                    typename std::iterator_traits<FwdIter>::value_type
               >
           )>
```

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp [hunk 5]
```diff
@@ -140,13 +448,60 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+#include <hpx/parallel/algorithms/remove.hpp>
+
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 {
+
+    template <typename ExPolicy, typename Rng, typename T,
+        typename Proj = util::projection_identity,
+        HPX_CONCEPT_REQUIRES_(
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp:490]: Could you disable clang-formatting for the requires clause, please?

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp [hunk 6]
```diff
@@ -162,3 +517,217 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+      : hpx::functional::tag<remove_if_t>
+    {
+    private:
+        // clang-format off
+        template <typename FwdIter, typename Sent, typename Pred,
+            typename Proj = hpx::parallel::util::projection_identity,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<FwdIter>::value &&
+                hpx::parallel::traits::is_projected<Proj, FwdIter>::value &&
+                hpx::traits::is_sentinel_for<Sent, FwdIter>::value
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp:537]: Please add a check for the predicate as well

**Comment by gonidelis** [comment]
@hkaiser Even though build-and-test timedout for some reason in the unit.test check, I think this is ok now.

**Review by hkaiser — APPROVED** [review]
@gonidelis yes, the failures are unrelated. Let's get this merged.

**Pull request merged**